custom/icds_reports/reports/poshan_progress_dashboard_data.py

```diff
@@ -0,0 +1,199 @@
+    _calculate_percent, _handle_average
+)
+from custom.icds_reports.utils import apply_exclude
+
+ICDS_CAS_COVERAGE_OVERVIEW = [
+    "Number of States Covered", "Number of Districts Covered", "Number of Blocks Covered",
+    "Number of AWCs Launched", "% Number of Days AWC Were opened", "% of Home Visits"]
+
+SERVICE_DELIVERY_OVERVIEW = [
+    "% of children between 3-6 years provided PSE for atleast 21+ days", "Weighing efficiency",
+    "% of trimester three women counselled on immediate and EBF", "Height Measurement Efficiency",
+    "% of children between 6 months -3 years, P&LW provided THR for atleast 21+ days",
+    "% of children between 3-6 years provided SNP for atleast 21+ days"]
+
+ICDS_CAS_COVERAGE_COMPARITIVE_MAPPING = {
+    "AWC Open": "avg_days_awc_open_percent",
+    "Home Visits": "visits_percent"
+}
+SERVICE_DELIVERY_COMPARITIVE_MAPPING = {
+    "Pre-school Education": "pse_attended_21_days_percent",
+    "Weighing efficiency": "weighed_percent",
+    "Height Measurement Efficiency": "height_measured_in_month_percent",
+    "Counselling": "counsel_immediate_bf_percent",
+    "Take Home Ration": "thr_percent",
+    "Supplementary Nutrition": "lunch_count_21_days_percent"
+}
+
+
+@icds_quickcache([
+    'domain', 'location_filters', 'year', 'month', 'step', 'quarter', 'include_test'
+], timeout=30 * 60)
+def get_poshan_progress_dashboard_data(domain, year, month, quarter, data_format, step, location_filters,
+                                       include_test=False):
+    def calculate_percentage_single_row(row):
+        for k, v in COLS_PERCENTAGE_RELATIONS.items():
+            num = row[v[0]]
+            den = row[v[1]]
+            extra_number = v[2] if len(v) > 2 else None
+            row[k] = _calculate_percent(num, den, extra_number)
+        return row
+
+    def calculate_aggregated_row(data, aggregation_level, data_format):
+        aggregated_row = {}
+        count = 0
+        for row in data:
+            count = count + 1  # count is used for number of states columns
+            for k, v in row.items():
+                if k not in aggregated_row.keys():
+                    aggregated_row[k] = 0
+                if k in ['state_name', 'district_name', 'district_id', 'state_id']:
+                    aggregated_row[k] = v
+                else:
+                    aggregated_row[k] += v if v else 0
+        # for quarter we need to average summation
+        if data_format == 'quarter':
+            for k, v in aggregated_row.items():
+                if k not in ['state_name', 'district_name', 'district_id', 'state_id']:
+                    aggregated_row[k] = _handle_average(v)
+        aggregated_row = calculate_percentage_single_row(deepcopy(aggregated_row))
+        aggregated_row = prepare_structure_aggregated_row(deepcopy(aggregated_row), count, aggregation_level)
+        return aggregated_row
+
+    def prepare_structure_aggregated_row(row, count, aggregation_level):
```

Comment:
good to pull all these fuctions out of the main function and make then protected to this file instead of making themnested functions

Reply:
updated